In [20]:
import numpy as np

In [21]:
def f2(x,y):
    return x**4 + y**4 - 2*(x-y)**2

In [22]:
Dx = lambda f,x,y,h=1e-5: (f(x+h,y) - f(x-h,y))/(2*h)
Dy = lambda f,x,y,h=1e-5: (f(x,y+h) - f(x,y-h))/(2*h)

In [23]:
x0, y0 = 0,3

In [24]:
Gradient = lambda f,x,y: np.array([Dx(f,x,y),Dy(f,x,y)])
Gradient(f2,x0,y0)

array([12., 96.])

In [25]:
def Momentum(x0, f, n, gamma = 0.001, N = 1000):
    r = np.zeros((N,2))
    r[0] = x0
    
    Grad = np.zeros((N,2))
    Grad[0] = Gradient(f,r[0,0],r[0,1])
    
    # We save the gradient in each step
    error = 10**(-5)
    it = 0

    for i in (range(1,N)):
        r[i] = r[i-1] - gamma*Gradient(f,r[i-1,0],r[i-1,1]) + n*(r[i-2] - r[i-3])
        Grad[i] = Gradient(f,r[i-1,0],r[i-1,1])
        it += 1
        if np.sqrt((Grad[i][0])**2 + (Grad[i][1])**2) < error:
            return r, it
    return r,it

In [26]:
x0=np.array([0,3])
n=0.6
R,it_normal = Momentum(x0,f2,n)
print("El número de iteraciones con mementum es: "+str(it_normal))


El número de iteraciones con mementum es: 342


In [27]:
R

array([[ 0.        ,  3.        ],
       [-0.012     ,  2.904     ],
       [-0.02366399,  4.61770376],
       ...,
       [ 0.        ,  0.        ],
       [ 0.        ,  0.        ],
       [ 0.        ,  0.        ]])

In [28]:
for n in np.arange(0, 1.01, 0.01):
    r, it = Momentum(x0,f2,n)
    print(n,it)
    if it == it_normal/2:
        it_mitad = it
        print(f"Sí existe un valor de coeficiente de momentum para el cual el algoritmo mejorado converga el doble de rapido y es {n}")
        break
    elif n == 1.0:
        print("No existe un valor de coeficiente de momentum para el cual el algoritmo mejorado converga el doble de rapido exáctamente")

0.0 975
0.01 964
0.02 953
0.03 943


0.04 932
0.05 922
0.06 911
0.07 901
0.08 890
0.09 880
0.1 869
0.11 859
0.12 848
0.13 838
0.14 827
0.15 817
0.16 806
0.17 796
0.18 785
0.19 775
0.2 764
0.21 754
0.22 743
0.23 733
0.24 722
0.25 712
0.26 701
0.27 691
0.28 681
0.29 670
0.3 660
0.31 649
0.32 639
0.33 628
0.34 618
0.35000000000000003 607
0.36 597
0.37 586
0.38 576
0.39 565
0.4 555
0.41000000000000003 544
0.42 534
0.43 523
0.44 513
0.45 502
0.46 492
0.47000000000000003 481
0.48 470
0.49 460
0.5 449
0.51 439
0.52 428
0.53 417
0.54 407
0.55 396
0.56 385
0.5700000000000001 375
0.58 364
0.59 353
0.6 342
0.61 331
0.62 320
0.63 309
0.64 298
0.65 287
0.66 275
0.67 262
0.68 246
0.6900000000000001 214
0.7000000000000001 245
0.71 245
0.72 285
0.73 315
0.74 379
0.75 396
0.76 349
0.77 351
0.78 340
0.79 331
0.8 353
0.81 352
0.8200000000000001 388
0.8300000000000001 456
0.84 365
0.85 371
0.86 386
0.87 374
0.88 424
0.89 475
0.9 560
0.91 625
0.92 664
0.93 735
0.9400000000000001 974
0.9500000000000001 999
0.96 999
0.97 999
0.98 151
0.99 81
1.

Sin embargo, al calcular el minimizer sin momentum, el número de iteraciones es 975. Esto quiere decir que, sin bien no se encontró un valor de coeficiente de momentum para el cual el algoritmo mejorado converga el doble de rapido exáctamente, si existe un valor de momentum para el cual el algoritmo mejorado converga MÁS del doble de rapido, y ese valor de momentum es 0.6